Feature: Added advanced options when creating archives

cinqmilleans opened the pull request:
**Resolved / Related Issues**
This pr improves the creation of compressed archives. Closes #10327.
1) The dialog has been redone and adds additional possibilities. I tried to stay practical and not technical.
2) The context menu has been unified with the decompression and offers additional possibilities.
3) There is no more distinction of the name of the archive according to whether one compresses an element or several.
4) The progress percentage indication is updated at each percent. Previously, it could never be updated for a large folder.
5) The compression code has been moved and is no longer static. It should be moved to the backend but the library needs to be updated first.

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

**Screenshots (optional)**
![Archive3](https://user-images.githubusercontent.com/46631671/202915201-c6331598-ec0f-4204-8a1e-0b92db74e5d4.png)
![Archive4](https://user-images.githubusercontent.com/46631671/202915202-8dc2174b-aec1-4f70-88be-f72041ef5f14.png)
![Archive5](https://user-images.githubusercontent.com/46631671/202915203-4e2598bd-ce94-4be8-872b-d26a86217241.png)
![Archive1](https://user-images.githubusercontent.com/46631671/202915197-f8d84611-ca32-4ed9-b1f2-565cbc912942.png)
![Archive2](https://user-images.githubusercontent.com/46631671/202915200-4bad56cb-5837-42d2-8e57-84429bbf9891.png)

yaira2 commented:
![image](https://user-images.githubusercontent.com/39923744/202915997-39a37478-3cf2-46d7-89b6-4e08a9b57fa1.png)

I have a couple of questions,
1. Is there a reason for having compress when selecting an existing archive?
2. Can the split options be hidden when they aren't available?
3. Icons in the dialog aren't necessary, can they be removed?
4. What does the "Do not compress" option do?
5. Using `x:Load` instead of `Visibility` is a good idea and will automatically fix the extra space at the bottom of the dialog.

yaira2 commented on src/Files.App/Dialogs/CompressArchiveDialog.xaml:
```diff
@@ -0,0 +1,169 @@
+﻿<ContentDialog
+	x:Class="Files.App.Dialogs.CompressArchiveDialog"
```
Should this be called "CreateArchiveDialog"?

cinqmilleans replied:
It's done.

cinqmilleans commented:
1) I changed the rule. The Compression part no longer appears when the selection is a single file which is an archive.
2) It's changed.
3) Icons have been removed.
4) There are 6 levels of compression (Ultra, High, Normal, Low, Fast, None). Currently, the compression is always in Ultra. It's handy to have None because it's very fast to store many files in one to transfer or store. Media files or documents are often already compressed but recompressing them is slow and unnecessary. I put back the 6 levels because, after testing, the Ultra level is very slow (40 minutes for my extreme test, against 4 minutes in Normal). The default mode, including without going through the dialog, is Normal because it is the best compression/speed compromise. We can optionally not display each level for simplicity (Ultra/Normal/Fast/None).
5) The extra space was intentional to avoid changing the window size. I changed the behavior to no longer have that empty space. I applied x:Load.

![Archive6](https://user-images.githubusercontent.com/46631671/203367587-48d787e5-980c-43b5-8bec-b77d4211adfe.png)
![Archive7](https://user-images.githubusercontent.com/46631671/203367591-03f789f7-214e-4c0c-9a8d-710a5c8ae6c9.png)

cinqmilleans commented on src/Files.App/Dialogs/CreateArchiveDialog.xaml:
```diff
@@ -0,0 +1,164 @@
+		</Grid>
+
+		<StackPanel Spacing="4">
+			<Grid Style="{StaticResource GridSettingStyle}">
```
This grid is used to display the background of each setting, as in the UserControl SettingsBlockControl. Using the style on the UserControl does not work.

yaira2 commented on src/Files.App/Dialogs/CreateArchiveDialog.xaml:
```diff
@@ -0,0 +1,164 @@
+			<Grid
+				x:Name="SplittingSizeLayout"
+				Style="{StaticResource GridSettingStyle}"
+				Visibility="Collapsed">
```
Let's switch to `x:Load` and bind to a property in the ViewModel.

cinqmilleans replied:
It's done

yaira2 commented on src/Files.App/Strings/en-US/Resources.resw:
```diff
@@ -2859,4 +2859,76 @@
+    <value>Archive</value>
+  </data>
+  <data name="CreateArchive" xml:space="preserve">
+    <value>Create a archive</value>
```
```suggestion
    <value>Create archive</value>
```

cinqmilleans replied:
It's done.

hez2010 commented:
Will do the review tomorrow. I'm running out of bandwidth today :)

cinqmilleans commented:
A few minor edits.
1) Opening the expander now automatically gives focus to the password field.
2) Combobox are too short for SplittingSize texts. To correct this problem, the combobox no longer displays the media when it is closed. The size becomes a minimum size because the "Do not split" text will be longer in many languages and it was truncated.

Other available formats are GZip, BZip2, Tar and XZ. I don't really know these formats but only Tar is managed by Files in decompression.

yaira2 commented on src/Files.App/Dialogs/CreateArchiveDialog.xaml:
```diff
@@ -0,0 +1,160 @@
+				x:Name="FileFormatSelector"
+				Grid.Row="0"
+				Grid.Column="1"
+				MinWidth="140"
```
Can you hardcode a width instead? `160` is probably enough.

cinqmilleans replied:
It's done.

yaira2 commented on src/Files.App/Dialogs/CreateArchiveDialog.xaml:
```diff
@@ -0,0 +1,160 @@
+						<StackPanel Orientation="Horizontal" Spacing="4">
+							<TextBlock Text="{Binding Label}" />
+							<TextBlock Text="{Binding Separator}" Visibility="{Binding IsDropDownOpen, ElementName=SplittingSizeSelector}" />
+							<TextBlock Text="{Binding Description}" Visibility="{Binding IsDropDownOpen, ElementName=SplittingSizeSelector}" />
```
Is description really needed? Files supports all these formats, so it seems a little redundant.

cinqmilleans replied:
The description concerns the type of media (cd, dvd, ...), not the format. I used a description to avoid the text being too long when closed.

![Archive8](https://user-images.githubusercontent.com/46631671/203837925-7dd20c28-5293-4c68-93a1-e3f3cbb415af.png)
![Archive9](https://user-images.githubusercontent.com/46631671/203837929-665bfbc5-32e3-40da-8ee0-7bce30f39e72.png)

yaira2 commented:
> Other available formats are GZip, BZip2, Tar and XZ. I don't really know these formats but only Tar is managed by Files in decompression.

Can we add support for rar and tar?

cinqmilleans commented:
You can add tar easily. For rar, you need another library, but also check the licenses. Free compression/decompression softwares doesn't seem to support it.

yaira2 commented:
Let's support whatever the current library supports 👍

hez2010 commented on src/Files.App/Dialogs/CreateArchiveDialog.xaml:
```diff
@@ -0,0 +1,160 @@
+				<ComboBox.ItemTemplate>
+					<DataTemplate>
+						<StackPanel Orientation="Horizontal" Spacing="4">
+							<TextBlock Text="{Binding Label}" />
+							<TextBlock Text="{Binding Separator}" Visibility="{Binding IsDropDownOpen, ElementName=SplittingSizeSelector}" />
+							<TextBlock Text="{Binding Description}" Visibility="{Binding IsDropDownOpen, ElementName=SplittingSizeSelector}" />
```
Can you use `x:Bind` instead? You can set `x:DataType` on `DateTemplate` to enable `x:Bind` here.

cinqmilleans replied:
I don't use DataType because the type is not public but nested to the dialog. This class does not need to be public as it is an implementation detail of the dialog. It is therefore not possible to use this type as a datatype in xaml.

hez2010 replied:
XAML doesn't recognize nested classes, but the class can be non-public. You can move the class out of the contained class and use x:DataType. We did the same for other classes a while ago.

cinqmilleans replied:
It's done.

hez2010 commented on src/Files.App/Files.App.csproj:
```diff
@@ -66,6 +66,9 @@
     </Content>
     <Content Remove="Assets\FilesHome.png" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="Dialogs\CreateArchiveDialog.xaml" />
+  </ItemGroup>
```
It's redundant.

cinqmilleans replied:
It's removed.

hez2010 commented on src/Files.App/Files.App.csproj:
```diff
@@ -120,4 +123,9 @@
     <TrimmerRootAssembly Include="CommunityToolkit.WinUI.UI.Controls.Media" />
     <TrimmerRootAssembly Include="CommunityToolkit.WinUI.UI.Controls.Primitives" />
   </ItemGroup>
+  <ItemGroup>
+    <Page Update="Dialogs\CreateArchiveDialog.xaml">
+      <XamlRuntime>$(DefaultXamlRuntime)</XamlRuntime>
+    </Page>
+  </ItemGroup>
```
It's redundant.

cinqmilleans replied:
It's removed.

hez2010 commented on src/Files.App/Filesystem/Archive/ArchiveCreator.cs:
```diff
@@ -0,0 +1,119 @@
+			try
+			{
+				var files = sources.Where(source => File.Exists(source)).ToArray();
+				var directories = sources.Where(source => System.IO.Directory.Exists(source)).ToArray();
```
```suggestion
				var directories = sources.Where(source => System.IO.Directory.Exists(source));
```

cinqmilleans replied:
It's done.

hez2010 commented on src/Files.App/Interacts/BaseLayoutCommandImplementationModel.cs:
```diff
@@ -576,52 +577,120 @@ public async Task CreateFolderWithSelection(RoutedEventArgs e)
+				return (sources, string.Empty, string.Empty);
+
+			string directory = associatedInstance.FilesystemViewModel.WorkingDirectory.Normalize();
+			string fileName = Path.GetFileName(sources.Length is 1 ? sources[0] : directory);
```
nit:

```suggestion
			string fileName = Path.GetFileName(sources is [source] ? source : directory);
```

cinqmilleans replied:
This does not compile as the feature is in preview. Maybe I need to activate something.

hez2010 commented on src/Files.App/Interacts/IBaseLayoutCommandImplementationModel.cs:
```diff
@@ -98,6 +98,10 @@ public interface IBaseLayoutCommandImplementationModel : IDisposable
 
 		Task CompressIntoArchive();
 
+		Task CompressIntoZip();
+
+		Task CompressIntoSevenZip();
+
```
Can we merge them into `CompressIntoArchive` which uses a parameter to decide the archive format?

cinqmilleans replied:
I prefer to use multiple methods because the role of this parameter would not be intuitive. CompressIntoArchive displays a dialog with several options. CompressIntoZip directly creates the archive without any options. Optionally, we can group CompressIntoZip and CompressIntoSevenZip together. But these functions should correspond to methods (see below).

hez2010 commented on src/Files.App/Interacts/BaseLayoutCommandsViewModel.cs:
```diff
@@ -174,6 +176,10 @@ private void InitializeCommands()
 
 		public ICommand CompressIntoArchiveCommand { get; private set; }
 
+		public ICommand CompressIntoZipCommand { get; private set; }
+
+		public ICommand CompressIntoSevenZipCommand { get; private set; }
+
```
Can we merge them into `CompressIntoArchiveCommand `?

cinqmilleans replied:
A command should never use a parameter. In particular this case which concerns very different behaviors. Using commands without parameters is needed to customize hotkeys (another feature in progress).

cinqmilleans commented:
I added support for other formats for compression. These formats do not support splitting. I will try to add them also in unzip (including browse content). This will be done in another pr.

cinqmilleans commented:
I disabled the other formats. bzip2 and xz don't seem to keep filenames. gz and tar do not support passwords. To decide what to do.

hez2010 commented on src/Files.App/Dialogs/CreateArchiveDialog.xaml:
```diff
@@ -0,0 +1,161 @@
+				SelectedValuePath="Key">
+				<ComboBox.ItemTemplate>
+					<DataTemplate>
+						<TextBlock Text="{Binding Label}" />
```
Still one left over here.

cinqmilleans replied:
This template was not necessary. I removed it. Thank you for your reviews. It's very nice of you.

hez2010 reviewed (approved):
LGTM. Thank you!
